### DISCUSSION TITLE
Symfony VarDumper Cloner tryes to clone array of traces and dies on getting reflection of "class_exists" function

### DISCUSSION THREAD

**Discussion by OleksiiBulba** [symfony/symfony#46709] [Q&A] [answered]
**P.S. See update at the end of the question.**

Hello, recently I have run my old local Symfony project (it was created several months ago with 6.0 version), I have updated it to Symfony 6.1 and the Web Debug Toolbar has been broken: I cannot get a response because token does not exists.

I have debugged the full flow of profiler, how it creates x-debug-token and then, when the request is terminated, it saves profile to var/cache/profiler folder.

The reason, why does not exist is: when profiler runs collect method on all collectors, logger collector clones traces, and then Cloner tries to get reflection on a function in the trace. The issue is that the Cloner does not recognizes "class_exists" function as a class and passes to Reflection class array with two empty arrays. In this moment my php script dies without any life signs.

A little bit of code for more understanding:
During the request, there is a deprecation log
```
The "Symfony\Bridge\Doctrine\Logger\DbalLogger" class implements "Doctrine\DBAL\Logging\SQLLogger" that is deprecated Use {@see \Doctrine\DBAL\Logging\Middleware} or implement {@see \Doctrine\DBAL\Driver\Middleware} instead.
```
with traces:
```
file="/var/www/html/vendor/symfony/config/Resource/ClassExistenceResource.php"
line={int} 78
function="class_exists"
```
When the Symfony dispatches KernelEvents::TERMINATE event, ProfileListener saves the profile to the cache folder:
```
// vendor/symfony/http-kernel/EventListener/ProfilerListener.php:121
// save profiles
foreach ($this->profiles as $request) {
    $this->profiler->saveProfile($this->profiles[$request]);
}
```
There is a foreach for running lateCollect method:
```
// vendor/symfony/http-kernel/Profiler/Profiler.php:96
        // late collect
        foreach ($profile->getCollectors() as $collector) {
            if ($collector instanceof LateDataCollectorInterface) {
                $collector->lateCollect();
            }
        }
```
and one of the collector is called `LoggerDataCollector` (`\Symfony\Component\HttpKernel\DataCollector\LoggerDataCollector`)
It collects some logger data:
```
// vendor/symfony/http-kernel/DataCollector/LoggerDataCollector.php:74
        if (isset($this->logger)) {
            $containerDeprecationLogs = $this->getContainerDeprecationLogs();
            $this->data = $this->computeErrorsCount($containerDeprecationLogs);
            // get compiler logs later (only when they are needed) to improve performance
            $this->data['compiler_logs'] = [];
            $this->data['compiler_logs_filepath'] = $this->containerPathPrefix.'Compiler.log';
            $this->data['logs'] = $this->sanitizeLogs(array_merge($this->logger->getLogs($this->currentRequest), $containerDeprecationLogs));
            $this->data = $this->cloneVar($this->data);
        }
```
and deep in Cloner class there is a code that tries to clone array with logger data, there are different cast methods etc.
At some point the cloner queue gets to the trace with `class_exists` function:
```
// vendor/symfony/var-dumper/Caster/ExceptionCaster.php:333
                $caller = isset($frame['class']) ? new \ReflectionMethod($frame['class'], $frame['function']) : new \ReflectionFunction($frame['function']);
                $stub->class .= ReflectionCaster::getSignature(ReflectionCaster::castFunctionAbstract($caller, [], $stub, true, Caster::EXCLUDE_VERBOSE));
```
Because the trace contains only a standard function `class_exists` without a class name, `$caller` is created from `ReflectionFunction` class. Next in `castFunctionAbstract`:
```
// vendor/symfony/var-dumper/Caster/ReflectionCaster.php:193
    public static function castFunctionAbstract(\ReflectionFunctionAbstract $c, array $a, Stub $stub, bool $isNested, int $filter = 0)
    {
        $prefix = Caster::PREFIX_VIRTUAL;

        self::addMap($a, $c, [
            'returnsReference' => 'returnsReference',
            'returnType' => 'getReturnType',
            'class' => 'getClosureScopeClass',
            'this' => 'getClosureThis',
        ]);

        if (isset($a[$prefix.'returnType'])) {
            $v = $a[$prefix.'returnType'];
            $v = $v instanceof \ReflectionNamedType ? $v->getName() : (string) $v;
            $a[$prefix.'returnType'] = new ClassStub($a[$prefix.'returnType'] instanceof \ReflectionNamedType && $a[$prefix.'returnType']->allowsNull() && 'mixed' !== $v ? '?'.$v : $v, [class_exists($v, false) || interface_exists($v, false) || trait_exists($v, false) ? $v : '', '']);
        }
```
Here addMap detects only a return type of the function, and as `$v` is `bool`, the second argument that is passed to `ClassStub` class is ["", ""] that corresponds to `$callable` option in that class:
```
// vendor/symfony/var-dumper/Caster/ClassStub.php:27
    public function __construct(string $identifier, callable|array|string $callable = null)
    {
        $this->value = $identifier;

        try {
            if (null !== $callable) {
                if ($callable instanceof \Closure) {
                    $r = new \ReflectionFunction($callable);
                } elseif (\is_object($callable)) {
                    $r = [$callable, '__invoke'];
                } elseif (\is_array($callable)) {
                    $r = $callable;
                } elseif (false !== $i = strpos($callable, '::')) {
                    $r = [substr($callable, 0, $i), substr($callable, 2 + $i)];
                } else {
                    $r = new \ReflectionFunction($callable);
                }
            } elseif (0 < $i = strpos($identifier, '::') ?: strpos($identifier, '->')) {
                $r = [substr($identifier, 0, $i), substr($identifier, 2 + $i)];
            } else {
                $r = new \ReflectionClass($identifier);
            }

            if (\is_array($r)) {
                try {
                    $r = new \ReflectionMethod($r[0], $r[1]);
                } catch (\ReflectionException) {
                    $r = new \ReflectionClass($r[0]);
                }
            }
```
as `$callable` is not a closure but an array, the class `ReflectionMethod` get two empty strings and php throws (maybe) a fatal and dies.
I tested this small part in 3v4l and `ReflectionMethod` throws fatal.

The question is: is there any issue with Cloner or I did not updated something correctly or should I just delete the project and start over? (as I don't have any valuable code there)

**UPDATE:**
So, I have reinstall the Symfony skeleton from the scratch, and the issue persists for Web Debug Toolbar. Now php dies during the creating Stub for `Symfony\Component\HttpKernel\EventListener\LocaleAwareListener::onKernelRequest` method when collecting called listeners on events.
As I understand the issue is with `vendor/symfony/var-dumper/Caster/ReflectionCaster.php:207` line because php dies inside it because of incorrect agruments.

**Comment by lyrixx** [comment]
Hi, It looks like there is an issue in Symfony if it occurs with a new projet, isn't?

Do you think you can create a reproducer?

**Reply by OleksiiBulba** [reply]
🤔 hmmm, could you help me with creating a reproducer?

**Reply by OleksiiBulba** [reply]
I have installed symfony 6.0 from scratch locally and there is no issue with Web Debug Toolbar 😥

**Reply by OleksiiBulba** [reply]
I have reached the place where php with Symfony 6.1 is dying but php with Symfony 6.0 is not. Here is 6.0 and `ReflectionMethod` has thrown an exception:
![image](https://user-images.githubusercontent.com/41155673/175123821-cfc9b19e-24a2-45bb-acf4-9cadcb01bd84.png)
but with Symfony 6.1 `ReflectionMethod` is not throwing an exception but rather kills php.
So possibly the issue might be not directly with Symfony itself, but with its performance (memory). E.g. I saw when Cloner starts to clone the variable, it turns off gc. I understand that Symfony 6.0 does the same, but my point is that maybe there is a memory leak somewhere...
My two empty projects from scratch differs only Symfony versions during the composer installation, php version, php ini are the same, xdebug version is the same. Even the list of installed Symfony components are the same (skeleton+maker-bundle+template (for twig)+profiler).

**Reply by lyrixx** [reply]
Thanks for investigating! Could you push your repository somewhere on GitHub?

**Reply by OleksiiBulba** [reply]
Hello @lyrixx, sorry for delay, I have prepared a [repository](https://github.com/OleksiiBulba/ScratchProject) with a scratch Symfony project (latest version) with an instruction in README.md where you can see how to run it. When you install it, just open localhost/home and you'll see that the Web Profiler Toolbar  cannot load itself. This is due to the fact that the main php script died after the page is loaded during the script termination.
Please do not hesitate to ask any question if you have any.

**Reply by OleksiiBulba** [reply]
Here how it looks like:
![image](https://user-images.githubusercontent.com/41155673/176006331-c8e5fa8d-8fdd-4ede-b6b4-266362e06fa7.png)

**Comment by OleksiiBulba** [comment, marked as answer]
I have upgraded from php 8.1.3 to 8.1.7 and the issue is gone. I suppose the root cause was in php: it somehow could not handle exception (maybe because of the memory or performance issues) and php died silently.
Now I created a new project from scratch with php 8.1.7 and the Web Profiler Toolbar was loaded correctly on the home page. The issue is resolved!

**Reply by lyrixx** [reply]
I'm glad it's fixed :)
